feat: task.run/run_async polymorphism support

pchalasani opened the pull request:
This PR supports `task.run/run_async` polymorphism, see this test: https://github.com/langroid/langroid/blob/task-run-polymorphism/tests/main/test_task_run_polymorphic.py

And also see an example of using this polymorphism to simplify a multi-agent search-critic system without inter-agent orchestration:
https://github.com/langroid/langroid/blob/task-run-polymorphism/examples/basic/multi-agent-search-critic-no-orch/main.py

The syntax is:
```
result = task.run(any_type_input, return_type=MyReturnType)
```

I looked into avoiding having to specify `return_type` and instead infer it from a result variable type-annotation, something like:

```
result: MyReturnType = task.run(some_input)
```

but it involves using unsafe things like inspecting the call frame.


@nilspalumbo Let me know if you have any ideas for enhancing etc
(No need to review the entire PR just the resulting end-user API, whether it can be enhanced, and if so, should this PR be merged and your enhancement can be on top of it)

nilspalumbo commented on langroid/agent/base.py:
```diff
@@ -1260,51 +1282,112 @@ def maybe_parse(tool: type[ToolMessage]) -> Optional[ToolMessage]:
+            if issubclass(list_element_type, ToolMessage):
+                # output_type = List[<ToolMessage],
+                # i.e. list of ToolMessage-derived objects
+                return cast(T, tools)
```
As in the next case, we might want to filter the tools by those which are instances of `output_type`

nilspalumbo replied:
Pushed this change, but the original may be more readable, let me know what you think

nilspalumbo replied:
Made this change

nilspalumbo commented on langroid/agent/task.py:
```diff
@@ -582,16 +587,44 @@ def reset_all_sub_tasks(self) -> None:
     def run(
         self,
-        msg: Optional[str | ChatDocument] = None,
+        msg: Any | None = None,
```
Here, and in `run_async`, this can just be `Any`

nilspalumbo replied:
Pushed this change, but the original may be more readable, let me know what you think

nilspalumbo commented:
Thanks! Looks great, I've added some changes which enable a cleaner interface; we can now say `task[return_type].run(...)`. Currently, the argument to `task.run()` overrides, and I've added a `default_return_type` argument to the `Task` constructor; would it be better to use the new interface exclusively?

pchalasani commented:
> Thanks! Looks great, I've added some changes which enable a cleaner interface; we can now say `task[return_type].run(...)`. Currently, the argument to `task.run()` overrides, and I've added a `default_return_type` argument to the `Task` constructor; would it be better to use the new interface exclusively?

the `__get_item__` idea is clever! I think we can keep the original `return_type`
Will merge once tests pass
